napari/layers/shapes/_shape_list.py

```diff
@@ -353,6 +353,178 @@ def add(
             # Set z_order
             self._update_z_order()
 
+    def add_multiple(
+        self,
+        shapes,
+        face_colors=None,
+        edge_colors=None,
+        z_refresh=True,
+    ):
+        """Add multiple shapes at once (faster than adding them one by one)
+
+        Parameters
+        ----------
+        shapes : iterable of Shape
+            Each Shape must be a subclass of Shape, one of "{'Line', 'Rectangle',
+            'Ellipse', 'Path', 'Polygon'}"
+        face_colors: iterable of face_color
+        edge_colors: iterable of edge_color
+        z_refresh : bool
+            If set to true, the mesh elements are reindexed with the new z order.
+            When shape_index is provided, z_refresh will be overwritten to false,
+            as the z indices will not change.
+            When adding a batch of shapes, set to false  and then call
+            ShapesList._update_z_order() once at the end.
+
+        TODO: Currently shares a lot of code with `add()`, with the
+        difference being that `add()` supports inserting shapes at a specific
+        `shape_index`, whereas `add_multiple` will append them as a full batch
+        """
+
+        all_z_index = list()
+        all_face_color = list()
+        all_edge_color = list()
+        all_vertices = list()
+        all_index = list()
+        all_mesh_vertices = list()
+        all_mesh_vertices_centers = list()
+        all_mesh_vertices_offsets = list()
+        all_mesh_vertices_index = list()
+        all_mesh_triangles = list()
+        all_mesh_triangles_index = list()
+        all_mesh_triangles_colors = list()
+
+        m_mesh_vertices_count = len(self._mesh.vertices)
+
+        for shape, face_color, edge_color in zip(
+            shapes, face_colors, edge_colors
+        ):
+
+            if not issubclass(type(shape), Shape):
+                raise ValueError(
+                    trans._(
+                        'shape must be subclass of Shape',
+                        deferred=True,
+                    )
+                )
+
+            shape_index = len(self.shapes)
+            self.shapes.append(shape)
+            all_z_index.append(shape.z_index)
+
+            if face_color is None:
+                face_color = np.array([1, 1, 1, 1])
+            all_face_color.append(face_color)
+
+            if edge_color is None:
+                edge_color = np.array([0, 0, 0, 1])
+            all_edge_color.append(edge_color)
+
+            all_vertices.append(shape.data_displayed)
+
+            index = [shape_index] * len(shape.data)
+            all_index.append(index)
+
+            # Add faces to mesh
+            m = m_mesh_vertices_count
+            vertices = shape._face_vertices
+            all_mesh_vertices.append(vertices)
+            m_mesh_vertices_count += len(vertices)
+            vertices = shape._face_vertices
```

Comment:
Optional: I realize this is from `add`, but consider removing many of the assignments to `vertices` because it's hard to reason about whether this actually affects code after this block (I think it doesn't).

In this case, you could just write `all_mesh_vertices_centers.append(shape._face_vertices)`, which seems fine to me. I think there are other cases below that could also be rewritten this way without an effect on behavior.